server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -6606,9 +6589,15 @@ private VMInstanceVO preVmStorageMigrationCheck(Long vmId) {
         }
 
         HypervisorType hypervisorType = vm.getHypervisorType();
-        if (vm.getType() != VirtualMachine.Type.User && !HYPERVISORS_THAT_CAN_DO_STORAGE_MIGRATION_ON_NON_USER_VMS.contains(hypervisorType)) {
-            throw new InvalidParameterValueException(String.format("Unable to migrate storage of non-user VMs for hypervisor [%s]. Operation only supported for the following"
-                    + " hypervisors: [%s].", hypervisorType, HYPERVISORS_THAT_CAN_DO_STORAGE_MIGRATION_ON_NON_USER_VMS));
+        if (vm.getType() != VirtualMachine.Type.User &&
+                (!hypervisorType.isFunctionalitySupported(HypervisorType.Functionality.VmStorageMigration)
+                        || hypervisorType.equals(HypervisorType.XenServer))) {
+
+            List<HypervisorType> supportedHypervisors = HypervisorType.getListOfHypervisorsSupportingFunctionality(HypervisorType.Functionality.VmStorageMigration)
+                    .stream().filter(hypervisor -> !hypervisor.equals(HypervisorType.XenServer)).collect(Collectors.toList());
```

Comment:
@vishesh92 seems better to add functionality VmStorageMigrationForNonUserVMs? (XenServer check is confusing, and chances of missing it wherever this is used later, so either move it to a method or add functionality)